src/main/kotlin/com/exactpro/th2/codec/util/MessageUtil.kt

```diff
@@ -14,40 +14,40 @@
  * limitations under the License.
  */
 
+@file:Suppress("unused")
+
 package com.exactpro.th2.codec.util
 
 import com.exactpro.th2.common.grpc.AnyMessage
 import com.exactpro.th2.common.grpc.AnyMessage.KindCase.MESSAGE
 import com.exactpro.th2.common.grpc.AnyMessage.KindCase.RAW_MESSAGE
+import com.exactpro.th2.common.grpc.EventID
+import com.exactpro.th2.common.grpc.Message
 import com.exactpro.th2.common.grpc.MessageGroup
 import com.exactpro.th2.common.grpc.MessageID
 import com.exactpro.th2.common.grpc.MessageMetadata
 import com.exactpro.th2.common.grpc.RawMessage
+import com.exactpro.th2.common.grpc.RawMessageMetadata
 import com.exactpro.th2.common.message.message
 import com.exactpro.th2.common.message.plusAssign
 import com.exactpro.th2.common.message.toJson
 import com.exactpro.th2.common.value.toValue
 
-val MessageGroup.parentEventId: String?
-    get() = messagesList.firstNotNullOfOrNull { anyMessage ->
-        when {
-            anyMessage.hasMessage() -> anyMessage.message.parentEventId.id.ifEmpty { null }
-            anyMessage.hasRawMessage() -> anyMessage.rawMessage.parentEventId.id.ifEmpty { null }
-            else -> null
-        }
+val AnyMessage.parentEventId: EventID?
+    get() = when {
+        hasMessage() -> with(message) { if (hasParentEventId()) parentEventId else null }
+        hasRawMessage() -> with(rawMessage) { if (hasParentEventId()) parentEventId else null }
+        else -> error("Unsupported $kindCase kind")
     }
 
+val MessageGroup.parentEventId: EventID?
+    get() = messagesList.firstNotNullOfOrNull { anyMessage -> anyMessage.parentEventId }
+
 /**
  * Returns parent event ids from each message.
  */
-val MessageGroup.allParentEventIds: Set<String>
-    get() = messagesList.mapNotNullTo(HashSet()) { anyMessage ->
-        when {
-            anyMessage.hasMessage() -> anyMessage.message.parentEventId.id.ifEmpty { null }
-            anyMessage.hasRawMessage() -> anyMessage.rawMessage.parentEventId.id.ifEmpty { null }
-            else -> null
-        }
-    }
+val MessageGroup.allParentEventIds: Set<EventID>
```

Comment:
I think we should allocate the set because a error can happen during traverse event ids and we should traverse again in error handler

Reply:
> I think we should allocate the set because a error can happen during traverse event ids and we should traverse again in error handler

Sequence can be traversed again (if it's not a constrained-once sequence ofc)